lib/imagery/sig.c

```diff
@@ -91,19 +123,87 @@ int I_read_one_signature(FILE * fd, struct Signature *S)
     return 1;
 }
 
+/*!
+ * \brief Read signatures from file
+ *
+ * File stream should be opened in advance by call to
+ * I_fopen_signature_file_old()
+ * It is up to caller to fclose the file stream afterwards.
+ *
+ * There is no need to initialize struct Signature in advance, as this
+ * function internally calls I_init_signatures.
+ *
+ * \param pointer to FILE*
+ * \param pointer to struct Signature *S
+ *
+ * \return 1 on success, -1 on failure
+ */
 int I_read_signatures(FILE * fd, struct Signature *S)
 {
-    int n;
+    int ver, n, pos;
+    char c, prev;
+    char bandref[GNAME_MAX];
 
+    I_init_signatures(S, 0);
     S->title[0] = 0;
-    while ((n = fgetc(fd)) != EOF)
-	if (n == '#')
-	    break;
-    if (n != '#')
-	return -1;
+    /* File of signatures must start with its version number */
+    if (fscanf(fd, "%d", &ver) != 1) {
+        G_warning(_("Invalid signature file"));
+        return -1;
+    }
+    /* Current version number is 1 */
+    if (ver != 1) {
+        G_warning(_("Invalid signature file version"));
+        return -1;
+    }
+
+    /* Goto title line and strip initial # */
+    while ((c = (char)fgetc(fd)) != EOF)
+        if (c == '#')
+            break;
     I_get_to_eol(S->title, sizeof(S->title), fd);
     G_strip(S->title);
 
+    /* Read band references and count them to set nbands */
+    n = 0;
+    pos = 0;
+    S->bandrefs = (char **)G_realloc(S->bandrefs, (n + 1) * sizeof(char **));
+    while ((c = (char)fgetc(fd)) != EOF) {
+        if (c == '\n') {
+            if (prev != ' ') {
+                bandref[pos] = '\0';
+                S->bandrefs[n] = G_store(bandref);
+                n++;
+            }
+            S->nbands = n;
+            break;
+        }
+        if (c == ' ') {
+            bandref[pos] = '\0';
+            S->bandrefs[n] = G_store(bandref);
+            n++;
+            /* [n] is 0 based thus: (n + 1) */
+            S->bandrefs = (char **)G_realloc(S->bandrefs, (n + 1) * sizeof(char **));
+            pos = 0;
+            prev = c;
+            continue;
+        }
+        /* Band references are limited to GNAME_MAX - 1 + \0 in length;
+         * n is 0-based */
+        if (pos == (GNAME_MAX - 2)) {
+            G_warning(_("Invalid signature file: band reference length limit exceeded"));
+            return -1;
+        }
+        bandref[pos] = c;
+        pos++;
+        prev = c;
+    }
+
+    if (! S->nbands > 0) {
```

Comment:
```suggestion
    if (!(S->nbands > 0)) {
```

clang issues: 
```
sig.c:202:9: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
    if (! S->nbands > 0) {
        ^           ~
```

Reply:
Thanks for checking these. We need to have to make existing GCC tests more strict and/or add clang tests to CI.

Reply:
> Thanks for checking these. We need to have to make existing GCC tests more strict and/or add clang tests to CI.

There is an issue regarding adding clang on CI with #1259. The PRs addressing warnings reported with #1247, have fixed all compiler warnings (but one) on my build system. Any warnings coming up now are newly introduced issues.

I did make an attempt addressing this with https://github.com/nilason/grass/tree/ci-test-clang but never got it working.

Reply:
clang in CI sounds good. In the meantime this issue is fixed.